File: research/data/polymarket_clob.py
"""Polymarket CLOB microstructure helpers (Python mirror).

Mirrors the pure-math portion of `src/tools/finance/polymarket-clob.ts`.
HTTP fetchers are intentionally omitted from the research package — live
network calls are TS-side only. Used by experiments to reason about
quality discounts identically to the production model.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Iterable, Sequence


@dataclass(frozen=True)
class ClobPricePoint:
    t_sec: int
    p: float


def parse_clob_price_history(raw: object) -> list[ClobPricePoint]:
    """Parse the `/prices-history` JSON shape into a sorted, valid series."""
    if not isinstance(raw, dict):
        return []
    hist = raw.get("history")
    if not isinstance(hist, list):
        return []
    out: list[ClobPricePoint] = []
    for item in hist:
        if not isinstance(item, dict):
            continue
        try:
            t = float(item.get("t"))
            p = float(item.get("p"))
        except (TypeError, ValueError):
            continue
        if t != t or p != p:  # NaN check
            continue
        if p < 0 or p > 1:
            continue
        out.append(ClobPricePoint(t_sec=int(t), p=p))
    out.sort(key=lambda x: x.t_sec)
    return out


def compute_price_velocity_pp_h(
    history: Sequence[ClobPricePoint], lookback_hours: int = 6
) -> float:
    """OLS slope of price (in pp) vs. time (in hours) over lookback window."""
    if len(history) < 2:
        return 0.0
    newest = history[-1].t_sec
    cutoff = newest - lookback_hours * 3600
    window = [pt for pt in history if pt.t_sec >= cutoff]
    if len(window) < 2:
        return 0.0

    n = len(window)
    sum_x = sum_y = sum_xy = sum_xx = 0.0
    for pt in window:
        x = pt.t_sec / 3600.0
        y = pt.p * 100.0
        sum_x += x
        sum_y += y
        sum_xy += x * y
        sum_xx += x * x
    denom = n * sum_xx - sum_x * sum_x
    if denom == 0:
        return 0.0
    return (n * sum_xy - sum_x * sum_y) / denom


def compute_max_hourly_jump(
    history: Sequence[ClobPricePoint], window_hours: int = 24
) -> float:
    """Maximum |Δp| between consecutive points within window_hours of latest."""
    if len(history) < 2:
        return 0.0
    newest = history[-1].t_sec
    cutoff = newest - window_hours * 3600
    max_abs = 0.0
    for i in range(1, len(history)):
        cur = history[i]
        prev = history[i - 1]
        if cur.t_sec < cutoff:
            continue
        d = abs(cur.p - prev.p)
        if d > max_abs:
            max_abs = d
    return max_abs

File: src/tools/finance/polymarket-clob.ts
/**
 * Polymarket CLOB API client + microstructure helpers.
 *
 * Implements the velocity / spike / spread fetches outlined in
 * docs/polymarket-prediction-improvements-research-2026-07.md §3.3 and §3.4.
 *
 * Pure-math helpers (`computePriceVelocityPpH`, `computeMaxHourlyJump`,
 * `parseClobPriceHistory`) are exported so they can be unit-tested without
 * live HTTP. The fetcher (`fetchClobSpread`, `fetchClobPriceHistory`) is
 * intentionally tiny — it is exercised in integration tests only.
 */

const CLOB_BASE = 'https://clob.polymarket.com';

export interface ClobPricePoint {
  tSec: number;
  p: number;
}

interface RawHistoryPoint {
  t?: unknown;
  p?: unknown;
}

interface RawHistoryResponse {
  history?: unknown;
}

/** Parse the `/prices-history` JSON shape into a sorted, finite, in-range series. */
export function parseClobPriceHistory(raw: unknown): ClobPricePoint[] {
  if (raw === null || typeof raw !== 'object') return [];
  const hist = (raw as RawHistoryResponse).history;
  if (!Array.isArray(hist)) return [];
  const out: ClobPricePoint[] = [];
  for (const item of hist as RawHistoryPoint[]) {
    if (item === null || typeof item !== 'object') continue;
    const t = Number(item.t);
    const p = Number(item.p);
    if (!Number.isFinite(t) || !Number.isFinite(p)) continue;
    if (p < 0 || p > 1) continue;
    out.push({ tSec: t, p });
  }
  out.sort((a, b) => a.tSec - b.tSec);
  return out;
}

/**
 * Linear-regression slope over the last `lookbackHours` of price history,
 * expressed in **percentage points per hour** (so a 0.01 → 0.04 ramp over
 * 3 hours yields ≈ 1.0, not 0.01).
 */
export function computePriceVelocityPpH(
  history: readonly ClobPricePoint[],
  lookbackHours = 6,
): number {
  if (history.length < 2) return 0;
  const newestSec = history[history.length - 1]!.tSec;
  const cutoff = newestSec - lookbackHours * 3600;
  const window = history.filter((pt) => pt.tSec >= cutoff);
  if (window.length < 2) return 0;
  // OLS slope of p (in pp = p * 100) vs. t (in hours).
  const n = window.length;
  let sumX = 0;
  let sumY = 0;
  let sumXY = 0;
  let sumXX = 0;
  for (const pt of window) {
    const x = pt.tSec / 3600;
    const y = pt.p * 100;
    sumX += x;
    sumY += y;
    sumXY += x * y;
    sumXX += x * x;
  }
  const denom = n * sumXX - sumX * sumX;
  if (denom === 0) return 0;
  return (n * sumXY - sumX * sumY) / denom;
}

/**
 * Maximum absolute hour-over-hour price change within the last `windowHours`.
 *
 * Used to detect whale prints / sharp regime shifts. Returned as a raw delta
 * in [0, 1] (i.e. 0.13 = 13 percentage-points).
 */
export function computeMaxHourlyJump(
  history: readonly ClobPricePoint[],
  windowHours = 24,
): number {
  if (history.length < 2) return 0;
  const newestSec = history[history.length - 1]!.tSec;
  const cutoff = newestSec - windowHours * 3600;
  let maxAbs = 0;
  for (let i = 1; i < history.length; i += 1) {
    const cur = history[i]!;
    const prev = history[i - 1]!;
    if (cur.tSec < cutoff) continue;
    const d = Math.abs(cur.p - prev.p);
    if (d > maxAbs) maxAbs = d;
  }
  return maxAbs;
}

// ---------------------------------------------------------------------------
// HTTP fetchers (kept thin; integration-tested only).
// ---------------------------------------------------------------------------

export interface ClobSpreadResponse {
  /** Spread in dollars on the YES token (e.g. 0.025 = 2.5pp). */
  spread: number;
}

export async function fetchClobSpread(tokenId: string): Promise<number | null> {
  try {
    const r = await fetch(`${CLOB_BASE}/spread/${encodeURIComponent(tokenId)}`);
    if (!r.ok) return null;
    const j = (await r.json()) as Partial<ClobSpreadResponse>;
    const s = Number(j?.spread);
    return Number.isFinite(s) && s >= 0 && s <= 1 ? s : null;
  } catch {
    return null;
  }
}

export async function fetchClobPriceHistory(
  market: string,
  interval: '1h' | '6h' | '1d' = '1h',
): Promise<ClobPricePoint[]> {
  try {
    const r = await fetch(
      `${CLOB_BASE}/prices-history?market=${encodeURIComponent(market)}&interval=${interval}`,
    );
    if (!r.ok) return [];
    return parseClobPriceHistory(await r.json());
  } catch {
    return [];
  }
}
